### ISSUE TITLE
Sledge Simplification Project

### ISSUE THREAD

**Issue by LogicAndTrick** [LogicAndTrick/sledge#238]
See issue #237 

The plan is to reduce project complexity so Sledge is a functional editor that's not too big for me to maintain.

- [ ] Upgrade to .NET 4.5 / remove Windows XP support
- [ ] Remove OpenGL 1 support and extra renderers
- [ ] Minimise features and functionality as much as possible
- [ ] Optimise resource loaders and messaging system to use async programming methods
- [ ] Other stuff here...

**Comment by LogicAndTrick** [comment]
I thought it would be a good idea to update this every now and then with progress. I'm making some fairly big changes so it will take a while to get to a stable release again. You can track my progress in the [simplification-project](https://github.com/LogicAndTrick/sledge/tree/feature/simplification-project) feature branch. 

## Current progress

### Main Window Rewrite

I'm replacing the main window with an empty shell. It still looks the same, but all the actual functional stuff will be moved into plugin projects. Examples of components:

- Commands - items in the menu bar and icon toolbar
- Sidebar items - items that show in the right/bottom sidebars
- Tools - items in the tool list on the left
- Documents - items that open in the main part of the app

All of these will have some way to be aware of "context" - for example some command might become active when a map is open, and one or more objects are selected. Certain actions will add/remove things from the state as they are opened and closed. Hopefully that removes dependencies from components that make everything difficult to work with.

### Async/Threading

I've written a basic WAD texture loader using an asynchronous pattern and it's working quite well. It will take a bit of work to port all the texture loaders so that will come later. Before that I also want to make async loaders for RMF/VMF/FGD/etc to make everything nice and threaded.

One long-term goal is to get the texture loader to watch for changes and reload the textures. For example if I add a texture to a WAD and save it, the texture will automatically update in the editor. Hopefully something like that isn't too difficult to do.

## Upcoming Plans

### Plugins

The plan is to have plugin support in the first release. All the standard components and tools will be built as plugins to make sure plugin creators have full control. I've been testing the [MEF](https://msdn.microsoft.com/en-us/library/dd460648(v=vs.110).aspx) framework for plugins and extensibility and it's going pretty well so far.

### Settings

I want to move settings to each plugin. Right now Sledge has way too many settings and it makes things difficult to maintain. The settings window is difficult to manage as well, so each plugin will have some way to publish a list of settings and there will be a generic way to manage them. 

### Multiple document types

I want to abstract out the tabs and documents system to allow multiple types of things to be opened. The most important one is the map document type, which is the only one that Sledge supports right now. But in theory, there could be another plugin that allows, for example, a WAD or MDL editor.

**Comment by LogicAndTrick** [comment]
Still making good progress but it's still fairly slow because there's quite a lot of work to be done. This month's update:

## Current Progress

![Sledge Shell screenshot](https://i.imgur.com/JM9gi6m.png)

### Main Window Rewrite

The shell for the main window is getting pretty close to being complete. I've got all the key components now: menu, sidebars, tools, commands, and hotkeys. The setup should make it very easy to add new components via plugins. Everything's using async as well, so it's much faster to load maps without having to wait for WADs to open (of course there's no textures for until they are fully loaded). Each component manages its own settings using a standard interface. I'll make a settings window later so users can edit them, but for now it's just JSON files in the AppData folder.

### Extensible Map Objects

Continuing the theme of plugins, the new map object (i.e. Solid, Entity, Face, etc.) system is very generic. Plugin makers will be able to add their own objects (such as curves which are specific to Quake 3) or special objects that only Sledge knows about. New objects won't be useful unless the plugin author also adds the proper classes to render, modify, save, and load those types. But all that should be possible to do.

### Rendering

It was thankfully very easy to plug the existing renderer into the new shell. I'm a lot happier with the way it's set up now. There's a few changes that need to be made when windows are created to make it more plugin-friendly, but I'll do that later. Texture loading code is a little bit cleaner now as well.

## Upcoming Plans

### Map Modification

Actually editing things is next on the list. I'll start with something easy like porting the brush creation tool to the new project. I'm hoping it shouldn't be too much effort to move all the existing code across without needing too many modifications. There's a lot of stuff to be done to make it all fit together again, though.

### Commands

Right most commands are all registered in the `DocumentSubscriptions` class, and it's a bit of a mess. I'm going to move them all so they have a class each. Plugins will work the same way.

### Translations

I've not started this yet, but I'm planning to get all UI strings to be loaded from an external file (XML or something). That should enable translations / multi-language support.

**Comment by Corruptinator** [comment]
Greetings! I found out about this open source software's existence by random chance through surfing the web, It's cool to see that there is a simplified open-source alternative that can edit goldsrc maps and others. I am looking forward to see support for the source engine to edit Half-life 2, Gmod, L4D2, or TF2 maps!  :)

P.S. I also found out that you can also import 3d object such as OBJ into Sledge, which is awesome! I am a 3d modeler on Blender and I have no problems in modeling low poly levels or stages. Plus I am aware that this is an _Alpha_ but could there be support for Mac and Linux one day? I know that sounds tedious to ask but it doesn't hurt to be curious.

**Comment by LogicAndTrick** [comment]
Glad to hear you're interested! Unfortunately the UI libraries for .NET aren't very good outside of Windows, especially when you need to use OpenGL. I've tried a few times before and it's far too difficult/unstable, so Linux/Mac support is not something I plan to spend any time on in the near future. If a stable, multi-platform UI framework for .NET with 3D support pops up at some point in the future I'll reconsider, but for now, there's no real option.

**Comment by Corruptinator** [comment]
Understood, sometimes things just dont work, but one can hope.  :)

**Comment by kinsi55** [comment]
I've just randomly stumbled upon this and found issue #207 from over two years ago. By no means am i trying to be rude, neither am i unaware of the complexity of such a project but i am wondering if / when we can expect support for Source games like CS:GO?

Its quite unfortunate that Valve has abandoned Hammer from improvements, mainly regarding the missing hardware accell.

**Comment by LogicAndTrick** [comment]
I doubt Sledge will support CSGO before it gets ported to Source 2. The Source 2 tools are a massive improvement over the Source SDK so you should be able to enjoy using modern tools once that happens.

Source will be supported eventually, but I don't know how long it will take to be stable. I've only recently started working on a fairly big rewrite after a long break, so I need to get it back to where I started before I can get back to adding Source support.

I'm hoping that I can get some Source support in this year, but I think I say that every year so please don't put too much weight on it :P

**Comment by kinsi55** [comment]
Honestly, i doubt a full Source 2 port will ever happen. Valve wouldnt risk this. All pretty much confirmed was that they will port the UI to Panorama, but porting **everything** to source 2 would mean a new physics engine (Rubikon vs Havok) which would completely fuck over movement / grenades / physics likely to turn away veterans. Also it would destroy the modding community. Essentially a full source 2 port would reset CS:GO back to 0 / bring it back a lot, and they have no intention to risk that seeing as it shows no sign of dying.

**Comment by LogicAndTrick** [comment]
You might be right, but they did it with DOTA without (major) issues so you never know! If Valve want to do something they'll just do it, they don't seem to care too much about community backlash. But if CSGO stays on Source, then Sledge will support it once Source support is added.

**Comment by kinsi55** [comment]
Yeah they did it with DOTA, but that really doesnt have anything that can break. Theres no Mods / modding community, neither are physics really important there. DOTA really could only improve from the port, so they did it. Looking forward to the future :D Thanks for maintaining this and the effort behind it, unfortunately my .net is pretty rusty so i wont be abel to help.

**Comment by LogicAndTrick** [comment]
Update: I've updated the first comment with the stuff that I have on my to-do list.

![image](https://cloud.githubusercontent.com/assets/3071180/25843850/05d98b86-34ec-11e7-8700-ce83612f87fa.png)

(Please excuse the crappy partial translation, it's just a test...)

Current progress:

- Translations are working! Translation files look like [this](https://github.com/LogicAndTrick/sledge/blob/feature/simplification-project/Sledge.Shell/Translations/Sledge.Shell.en.json) and [this](https://github.com/LogicAndTrick/sledge/blob/feature/simplification-project/Sledge.Shell/Translations/Sledge.Shell.jp.json).
- Select, brush, camera tools all (mostly) ported to new system
- Texture browser moved over (will likely leave out the 'favourite textures' section for now)
- Added status bar stuff to the main shell
- Added a generic interface for map elements to be serialised and deserialised
- I'm starting to feel like a Java programmer with all the interfaces and factories I'm making. It's definitely the best method for Sledge that I know of, though. Everything's fitting together really well.
- I added custom grids, something I've wanted to add for a while now. Only the "invisible" grid and the normal square grid are available right now, but it makes it easy to add other grids such as angled grids, hexagons, triangles, radial, etc.

Future plans:

- Porting the entity tool is next, and then the texture tool (which is the second most complex tool)
- Still haven't started on undo/redo, but I'm hoping it's pretty easy to slot in
- I'm thinking about a custom map format native to Sledge. It'd be the same text format as VMF, but in Sledge layout rather than Hammer layout. VMF will still be supported. RMF/JMF/MAP/etc aren't extensible formats so they can't support custom objects from plugins, so custom objects will need to convert themselves into primitive objects to save to those formats.

**Comment by SharpOB** [comment]
I doubt CS:GO will be ported for the same reason TF2 won't be: Physics matters alot. In CS:GO, you have grenade throws, and in TF2, you have rocket and sticky jumps as well as the flamethrower flames.

**Comment by SharpOB** [comment]
Also, you should definitely try running Sledge under Wine with Ubuntu 17.04 in VirtualBox. If it works, great.

**Comment by petrikov** [comment]
>try running Sledge under Wine with Ubuntu 17.04 in VirtualBox.

Geez, that's an overkill.

But in all seriousness, I'm going to give this a try a couple of weeks later. Hammer has some texture issues on Wine, at least that's what I got the last time I was running it on Wine. Plus it was Wine 2.0 instead of the latest development release.

**Comment by LogicAndTrick** [comment]
Next update: I've been taking a bit of a break this last week, but next week I'm hoping I can get a lot of stuff done. There's lots of progress since the last update though:

- Texture tool is ported - it was a lot easier than I expected! Very happy with it now.
- Entity, clip, and decal tools are ported as well.
- Undo/redo has been added, it was easy to add because I knew what was required from the beginning. When I first did undo/redo stuff, it was quite painful to implement because it was the first time I had ever done it. This time around, it was much easier.
- The settings editor is coming along very nicely, settings can be edited and saved, plus game configs and hotkey setups.
- Visgroups have been implemented
- Group/ungroup implemented 
- Various other things

Next things to work on are:

- VM tool - I've been avoiding this one because VM is painful, but I've gotta get this one ported over.
- Model rendering - difficult and annoying, I've been avoiding this for a while now...
- Cordon tool - fairly easy port, also need to implement cordon visibility as well.
- Compiling - probably a large part of this will just be copy/paste the same code files, so just lots of boring work for this.
- Save/load extra formats - I need to port the VMF/MAP parsers over, mostly mindless copy/paste for the most part.
- More commands and hotkeys - implement more editing commands and windows (e.g. transform, paste special, tie to entity, etc) - pretty easy stuff
- Some extra translation support, auto-update, and other housekeeping stuff

And that's getting very close to the complete list! There's a bunch of boring and time-consuming tasks in there that are pretty easy, but there's still some difficult tasks left as well. But it's getting closer to a test release! (Before anyone gets too excited, the test release will definitely have bugs, and no new Source features for now. This is all work I had to do in order to prepare for extra engines - but it's still pretty exciting if you ask me!)

**Comment by clintonbale** [comment]
Looking forward to the new release! Really want to try my hand at contributing to this project.

**Comment by LogicAndTrick** [comment]
Apologies for the lack of updates, I've been pretty busy with other stuff and haven't had a chance to make any progress. I'm hoping to get back into the swing of things this week!

**Comment by NCC-Lykos** [comment]
I'm so excited for plugins... we are currently working on Genesis Engine support in Sledge, but if it could be made as a plugin instead, that would be amazing.

**Comment by NukemD** [comment]
> I'm thinking about a custom map format native to Sledge. It'd be the same text format as VMF, but in Sledge layout rather than Hammer layout. VMF will still be supported. RMF/JMF/MAP/etc aren't extensible formats so they can't support custom objects from plugins, so custom objects will need to convert themselves into primitive objects to save to those formats.

Please DO NOT use the VMF format. It's a significant regression over the RMF format and causes severe issues with brush precision. Using the VMF format for HL1 mapping would be a disaster and I will recommend completely abandoning Sledge if this is implemented.

In Source Hammer, the brush precision is so bad that if you have a brush that isn't a perfect cuboid, it will get corrupted. Brushes are stored by clipped planes, which are defined by 3 points. These three points don't have enough decimal precision, so if you have a brush with obtuse or acute angles, it will result in the plane intersections between faces not matching up. So when Hammer tries to render the brush from this garbage data, in many cases multiple vertexes are created on brush corners that can be attached to random faces, or even completely orphan in some cases.

Here is an example of said brush corruption:
https://i.imgur.com/ud1pXoN.png

The vertex spam you see in the top right window was originally a single point. When the map is saved, the brush faces are corrupt and scatter around like this. These brushes have actually had faces separate, and you can see the result in the 3D view with the yellow cracks where the nodraw is showing through.

Unless you're explicitly working with a Source map, do not force VMF format use. RMF and MAP are far superior formats, despite their difficulties being worked with.

**Comment by SharpOB** [comment]
A sledge-only format would be great as long as maps were exportable to other formats, and that there are no issues with brush corruption

**Comment by LogicAndTrick** [comment]
> Please DO NOT use the VMF format. It's a significant regression over the RMF format and causes severe issues with brush precision. Using the VMF format for HL1 mapping would be a disaster and I will recommend completely abandoning Sledge if this is implemented.

First of all, Sledge will ALWAYS support as many formats as possible and I won't ever drop support for RMF or any others. Formats will only be added, not removed.

Secondly, Sledge VMFs are slightly different to Hammer VMFs in that they save vertex positions to the file rather than just the brush planes. This means that **you don't get brush corruption when using VMF with Sledge** (unless you use both Sledge + VHE on the same file). Open a Sledge VMF in notepad and you will see the difference. 

VMF is a much better format that RMF from a programming perspective because it's extensible. I can add Sledge-only sections to a VMF and it won't break Hammer support for the file. RMF will never have support for displacements, multiple visgroups, and so on because it's a binary format and any change to the format will break compatibility.

VMF is essentially a format similar to JSON or XML - you can add extra things and if a program doesn't support it, then it will just be ignored. Calling VMF a regression over RMF simply because VHE has a bug is ignorant and just plain incorrect.

And by the way, MAP is absolutely not a superior format. Not only does it have the exact same problem as VMF in that it doesn't save vertex positions, it also isn't extensible and it strips out all useful editor data such as colours, groups, visgroups, cameras, etc. Please don't use the MAP format, it's genuinely the worst option available and only exists because the compile tools use it.

**Comment by NukemD** [comment]
> Secondly, Sledge VMFs are slightly different to Hammer VMFs in that they save vertex positions to the file rather than just the brush planes. This means that you don't get brush corruption when using VMF with Sledge (unless you use both Sledge + VHE on the same file). Open a Sledge VMF in notepad and you will see the difference.

Creating a different version of the VMF format is a bad idea, it's just going to cause grief down the road. You contradict yourself later in your post where you say that adding extra features to the VMF won't cause problems, but the existing problem of brush corruption means that there is already a problem. Valve doesn't publicly acknowledge the problem, and few mappers even understand why it happens or how to avoid it, meaning there is already a compatibility issue.

If you think RMF is a bad format from a developer's perspective and want a format more like VMF, develop your own format and call it something else like SMF, that way people don't associate it with Hammer and destroy their maps.

> Calling VMF a regression over RMF simply because VHE has a bug is ignorant and just plain incorrect.

No, it's not ignorant because VMF _is a map format written specifically for Hammer._ It's not a universal file format like XML designed to be read by everyone.

> And by the way, MAP is absolutely not a superior format. Not only does it have the exact same problem as VMF in that it doesn't save vertex positions

I've been mapping for 20 years and Source mapping for 14 years, I rely on the MAP format to save me from the cancer that is VMF. I have never seen it corrupt one of my maps brushwork. Yes, it does have the downsides of not being able to save displacements and a few other features, but it beats losing weeks of work and having to chase down broken brushes that don't show up in any error checking algorithm.

Thankfully someone created a program called Hammerfix which injects itself into Hammer and stops the brush corruption from happening during map saves. It's an ugly hack, but I'm glad it exists, as it makes saving work far easier.

**Comment by LogicAndTrick** [comment]
Unfortunately you couldn't be more incorrect, VMF is a serialisation format just like XML and making a "custom" version that still works in other programs is trivial and I'm already doing it in Sledge. The rounding issues in Hammer's VMF code are a bug in Hammer. It's got nothing to do with the format.

If you open a VMF in notepad you will see that it uses the exact same syntax as VMT, resource files, and pretty much any other text-based format that the Source engine uses. It's XML with a slightly different syntax.

As I said RMF support will not be removed so feel free to keep using that format. Just don't spread misinformation about Sledge. It's Hammer that has the bug.

**Comment by SharpOB** [comment]
Make your own sledge format (even if it is similar to `.vmf`), but please do not use the `.vmf` extension. Make your own, such as `.smf`, because you might risk causing issues with Hammer.

**Comment by TeamSpen210** [comment]
There's no need for that. It's entirely possible to make it openable in both, although saving via Hammer would remove the extra features. Compression isn't really useful I don't think, you can already compress the files quite well with existing archive programs.

**Comment by LogicAndTrick** [comment]
Yep, exactly. As for compression, I've considered it but it makes it inconvenient for source control and it's easy to compress using some other method.

I still might add `.smf` or something because a backwards compatible VMF will be larger in size because it would have some duplicated data.

**Comment by Xeddius** [comment]
As someone who uses Linux as my main operating system, I'm saddened that there will likely not be support in the future. I have been thinking about starting to make maps again after a 10+ year hiatus, however, given that there's nothing modern that can run on my main computer, I'm probably not going to start modding again.

 Microsoft has released .net tools to make cross-platform a lot simpler, I don't know fi that helps, but I'm hopeful. Anyway, for those who do use Windows, I'm glad they're getting a more modern tool at least.

**Comment by Xeddius** [comment]
@FloraCanou I was not aware of JACK, thank you!
JACK is installed but required a lot of tweaking to get working (I'm on Arch and it looks like it was made for ubuntu.)
